### ISSUE TITLE
Thank you for your work,  could you teach me some code problems?

### ISSUE THREAD

**Issue by qyr0403** [bpiyush/TestOfTime#5]
In https://github.com/bpiyush/TestOfTime/blob/main/package/losses/weighted_contrastive.py
Line 38 and 39, I can't understand why do it perform "logits[batch_size // 2:, :batch_size // 2] += self.alpha":

`    def __call__(self, pooled_video, pooled_text, **kargs):
        batch_size = pooled_video.size(0)
        # change device of the weight
        self.loss.weight = self.loss.weight.to(pooled_video.device)
        self.alpha = self.alpha.to(pooled_video.device)
        logits = torch.mm(pooled_text, pooled_video.transpose(1, 0))
        logits[batch_size // 2:, :batch_size // 2] += self.alpha
        logits[:batch_size // 2, batch_size // 2:] += self.alpha
        targets = torch.arange(
            batch_size,
            dtype=torch.long,
            device=pooled_video.device)
        return self.loss(logits, targets)`

I know 'self.alpha' is parameter matrix about TNCE, but why do not multiply but add?
logits represent cosine similarity matrix about video and text, it should range from -1 to 1.

**Comment by bpiyush** [comment]
This is just an implementation trick. If you expand the loss term equation, you want something like this:

![image](https://github.com/bpiyush/TestOfTime/assets/19412343/ee3e6bec-5a0e-4d24-8999-90b66592baf5)

So you want the `alpha` multiplier to happen on the exponent. But the CrossEntropy loss module does the softmax within itself. So the trick is as follows: suppose you have any scalar $x$,

$$\alpha\exp(x) = \exp(\log(\alpha)).\exp(x) = \exp(\log(\alpha) + x)$$

So essentially, you need to add $\log(\alpha)$ to your logits $x$.

Hope this helps.

**Comment by qyr0403** [comment]
> This is just an implementation trick. If you expand the loss term equation, you want something like this:
> 
> ![image](https://private-user-images.githubusercontent.com/19412343/297780439-ee3e6bec-5a0e-4d24-8999-90b66592baf5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2NDY5MjYsIm5iZiI6MTcwNTY0NjYyNiwicGF0aCI6Ii8xOTQxMjM0My8yOTc3ODA0MzktZWUzZTZiZWMtNWEwZS00ZDI0LTg5OTktOTBiNjY1OTJiYWY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE5VDA2NDM0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2ZDVjMWVhNWZjYTllMjIyMTUzYzc4MzY3NGZmOGYzZDJjZDg2NzI4NjcxZDJkMzM3NDkxM2RiNDNjODc2MTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LVyImX5cdKCD8N-mhuq6OuCwQa2yAuQE6-o5yweGpjE)
> 
> So you want the `alpha` multiplier to happen on the exponent. But the CrossEntropy loss module does the softmax within itself. So the trick is as follows: suppose you have any scalar x,
> 
> αexp⁡(x)=exp⁡(log⁡(α)).exp⁡(x)=exp⁡(log⁡(α)+x)
> 
> So essentially, you need to add log⁡(α) to your logits x.
> 
> Hope this helps.

thank you very much.

**Comment by qyr0403** [comment]
thank you very much, i had not seen ''self.alpha = np.log(alpha + 1e-8)''
